### ISSUE TITLE
Mainclass not found after updating extension from 1.23.0 to 1.24.0

### ISSUE THREAD

**Issue by Thejipppp** [scalameta/metals#5512]
**Describe the bug**

After updating the Scala (metals) extension in VSCode on 2023/07/20, the mainclass is not found anymore and thus cannot run anymore in a project that uses a workspace. The workspace contains multiple sbt projects.
In a project without workspace, everything still works.
Opening only the project with the mainclass and running it, also still works, but then of course you don't see the subprojects anymore in the editor.

**To Reproduce** Steps to reproduce the behavior:

1. Have two sbt projects, the one with the mainclass depending on the other
2. Add a .vscode/myWorkspace.code-workspace file to the project with the mainclass
3. Link to the two projects in there `"folders": [
		{
			"path": ".."
		},
		{
			"path": "../../my-second-project"
		}
	]`
4. The mainclass is not found anymore when running the project in VSCode.

**Expected behavior**

The project runs as expected

**Installation:**

- Operating system: Windows
- VSCode version: 1.80.1
- VSCode extension version: 1.24.0
- Metals version: 1.0.0

**Search terms**

Visual Studio Code, Breaking Change, Update, Latest Version,

**Comment by kasiaMarek** [comment]
Yes, thanks. It seems this part was not adjusted properly for workspace folders. However, as @tgodzik said, the fact it worked before is accidental. Metals did not have support for workspace folders and always **only** the first folder was visible in metals.

**Comment by spangaer** [comment]
I don't think that's entirely true though.

While you needed to be smart about it, it did just work all this time. When using source dependencies in SBT, you just used the setting: `bloopAggregateSourceDependencies := true` to generate all bloop files in the consuming project. If you then made a multi-folder workspace in which the consuming project was the first listed folder. Everything just worked. E.g. all modules of both depending as depended project where visible in the metals project list.

It seems that with the multi-root support this somehow got broken. If you now started editing in the sourcedep project, metals spits out

```
2023.11.03 10:50:08 INFO  no build target found for <path to file>.scala. Using presentation compiler with project's scala-library version: 3.3.1
```

and the typing (and derived services like auto-complete) break entirely. It still shows the list of projects in the metals project list though.

The only way I seem to get it to work again is by opening VSCode in the parent folder and relying on project root detection.

I did like the old mode, because it had some cleanness to it. It also had the benefit that the launch.json configs in the different parent projects are available from within VSCode.

I'm also a bit worried that the reliance on project root detection is fragile. So I'm unsure if just converting our workspaces to that mode is the right thing to do right now?

**Comment by tgodzik** [comment]
> The only way I seem to get it to work again is by opening VSCode in the parent folder and relying on project root detection.

Next version will have a specific setting to hardcode your project root, which should be safer than automatic detection.

I understand the issue now and I am not sure if we can make the old behaviour work properly again. We could theoretically fall back to another folder for information if we didn't find it in the current one, but that might not be fast enough. The other option is to detect that a certain folder contains a target definition from another folder, which might be a difficult heuristic to make work always.

Anyway, might be good to raise a feature request for your use case

**Comment by spangaer** [comment]
Well at the very least I've created a reproducer. (was building it already)
https://github.com/spangaer/vscode-metals-folder-issue-reproducer

I was going to raise it as a separate issue, but if you tell it should be a feature request, I can do that too.
(if tell me which and where you prefer, I'll get to it)

**Comment by spangaer** [comment]
A workspace level setting that allows to choose whether to treat multiple roots as a single bloop build could do wonders too on that front.

I even used to not map in my sourcedep scala projects, but open it in a separate VSCode. But I use RedHat Java LSP for the Java parts and that one doesn't tolerate sourcedeps, which are not in single workspace.

**Comment by spangaer** [comment]
By writing previous I just figured pretty silly but effective workaround.

If I run `sbt bloopInstall`, both in the sourcedep and the depending project, the multi-root workspace sees them (similar with the 2 VSCodes before) as two independent projects. The sourcedep project being self-contained an the other one building the whole shabam.

Not entirely to surprise it does start working again like it did before. The annoying part is of course that means I always need to run the bloopInstall twice (or as my times as I have sourcedeps), because e.g. disabling that `bloopAggregateSourceDependencies` doesn't seem to do much.

If you wonder why I run bloopInstall manually, I've had struggles with bloop versions before (again with 1.5.10+ btw) and I need to pass some additional parameters that affect the bloop files.

**Comment by tgodzik** [comment]
> If you wonder why I run bloopInstall manually, I've had struggles with bloop versions before (again with 1.5.10+ btw) and I need to pass some additional parameters that affect the bloop files.

Did you already report that? I could take a look. Is it happening with the reproducer repository you posted?

Will the use case with custom project root work for you once released? I really don't see a better way since it's much harder to automatically make it work for your case. This is how I would see it:

1. Open the main directory.
2. Set the custom project root to the main (`metals.customProjectRoot`)
3. Import project

You can test it out already using a prerelease version. Looks to be working well for me.

**Comment by spangaer** [comment]
> Did you already report that? I could take a look. Is it happening with the reproducer repository you posted?

I have not gotten to that yet. Last time I ran in to problems it was solved with upgrading SBT. Which I haven't tried just yet. I also have some fear it's not that. I use a source dependency on a project, from the project build context of another project and in between those two there seems to be some cross-talk of coursier being turned off or on between those build levels. (I have coursier disabled, because it really doesn't play nice with local maven repo's). That messes up some dependency meta-data.

I still plan investigate and report tough.


> Will the use case with custom project root work for you once released? I really don't see a better way since it's much harder to automatically make it work for your case. This is how I would see it:
> 
>     1. Open the main directory.
> 
>     2. Set the custom project root to the main (`metals.customProjectRoot`)
> 
>     3. Import project
> 
> 
> You can test it out already using a prerelease version. Looks to be working well for me.

It's my understanding that would address the fragility of relying on the auto-detection, but of course that means that then my launch.json files are out of place. I sort of like that they work out-of-the-box for the depending and depended  project with the multi-folder workspace setup (both projects live an separate git repo's in practice, so are not tightly coupled).

But the workaround I figured after our prior discussion is working quite nice for me ATM, I have the experience I had before back, it probably works better even. E.g. things like find references used to be completely non-functional in the depended project and now at least they work within the project. 👍 

It comes at the price of the double `bloopInstall`, but I'm willing to take it right now. It's only when using the sourcedeps. Which is not the standard case.

Over time. there's chance that I'll be using double bloop or the `metals.customProjectRoot`, on a case-by-case basis. 🤷

**Comment by spangaer** [comment]
> > If you wonder why I run bloopInstall manually, I've had struggles with bloop versions before (again with 1.5.10+ btw) and I need to pass some additional parameters that affect the bloop files.
> 
> Did you already report that? I could take a look. Is it happening with the reproducer repository you posted?
> 

Somebody beat me to it, as this seems to be the bug:
https://github.com/sbt/sbt/issues/7434

**Comment by tgodzik** [comment]
Looks like a workaround is to use:
```
updateSbtClassifiers / useCoursier := true
```
does it help in your case?

**Comment by spangaer** [comment]
Seems to, indeed. Be it I added some sugar on top.

https://github.com/sbt/sbt/issues/5263#issuecomment-1852859135
